consensus/src/chained_bft/event_processor.rs

```diff
@@ -656,48 +680,44 @@ impl<T: Payload, P: ProposerInfo> EventProcessor<T, P> {
             return;
         }
 
+        self.add_vote(vote, quorum_size).await;
+    }
+
+    /// Add a vote. Fetch missing dependencies if required.
+    /// The `duplicates_expected` field is used for cases, in which some of the votes might
+    /// be duplicated (e.g., when the votes are attached to the timeout messages).
+    /// If a QC is formed then
+    /// 1) fetch missing dependencies if required, and then
+    /// 2) pass the new QC to the pacemaker, which can generate a new round in return.
+    async fn add_vote(&self, vote: VoteMsg, quorum_size: usize) {
         let deadline = self.pacemaker.current_round_deadline();
+        let preferred_peer = Some(vote.author());
+        let vote_round = vote.round();
         // TODO [Reconfiguration] Verify epoch of the vote message.
         // Add the vote and check whether it completes a new QC.
-        match self
-            .block_store
-            .insert_vote(vote.clone(), quorum_size)
-            .await
+        if let VoteReceptionResult::NewQuorumCertificate(qc) =
+            self.block_store.insert_vote(vote, quorum_size).await
         {
-            VoteReceptionResult::DuplicateVote => {
-                // This should not happen in general.
-                security_log(SecurityEvent::DuplicateConsensusVote)
-                    .error(VoteReceptionResult::DuplicateVote)
-                    .data(vote)
-                    .log();
-            }
-            VoteReceptionResult::NewQuorumCertificate(qc) => {
-                if self.block_store.need_fetch_for_quorum_cert(&qc) == NeedFetchResult::NeedFetch {
-                    if let Err(e) = self
-                        .sync_manager
-                        .fetch_quorum_cert(qc.as_ref().clone(), Some(vote.author()), deadline)
-                        .await
-                    {
-                        error!("Error syncing to qc {}: {:?}", qc, e);
-                        return;
-                    }
-                } else {
-                    if let Err(e) = self
-                        .block_store
-                        .insert_single_quorum_cert(qc.as_ref().clone())
-                        .await
-                    {
-                        error!("Error inserting qc {}: {:?}", qc, e);
-                        return;
-                    }
+            counters::TIMEOUT_VOTES_FORM_QC_COUNT.inc();
```

Comment:
How do we know this is timeout vote?

Reply:
Ugh, right, I moved it to a separate function that is now used in other places as well. Good catch, thanks! Let me fix it.